### DISCUSSION TITLE
Progress callback for initialization

### DISCUSSION THREAD

**Discussion by Irev-Dev** [donalffons/opencascade.js#20] [Ideas]
Not sure how easy it would be since this project heavily relies on codgen. An idea I had is having a progress callback for `initOpenCascade` for download progress of the wasm file. I think it would be useful since it can take a while to download depending on internet speed.

I know that `XMLHttpRequest` have [progress events](https://developer.mozilla.org/en-US/docs/Web/API/ProgressEvent)

Example usage:
```javascript
initOpenCascade((progress) => console.log(`loading: ${progress}%`)).then(opencascade => {
  // . . .
})
```

**Comment by donalffons** [comment]
A progress indicator certainly make a lot of sense in dealing with the long loading times. The [initOpenCascade](https://unpkg.com/opencascade.js@1.0.1/index.js) function is currently just a super simple wrapper around Emscripten's instantiation mechanism:
```js
import opencascade from "./dist/opencascade.wasm.js";
import wasmFile from "./dist/opencascade.wasm.wasm";

export const initOpenCascade = () => {
  return new opencascade({
    locateFile(path) {
      if(path.endsWith('.wasm')) {
        return wasmFile;
      }
      return path;
    }
  });
}
```
(This is necessary for supporting build systems like webpack, as the (potentially hashed) URL of the wasm file might not be not what Emscripten expects it to be). The actual fetching of the WASM files and instantiation is handled by Emscripten. Emscripten allows customization of that process with several properties of the Module Object. It seems like this function [`getPreloadedPackage(...)`](https://emscripten.org/docs/api_reference/module.html#Module.getPreloadedPackage) is the one that could be used to achieve a progress indicator. If you're implementing a custom `initOpenCascade` function in you project, you could probably achieve this already.

For the upcoming version, I think it would make sense to allow users to pass whatever prop they want to Emscripten's Module object / instantiation.

It would probably also make sense to have a simplified interface for progress reports as you suggested built-in to this library. But short-term, I think it is better if the number of interfaces maintained by this project stays as small as possible.

**Reply by Irev-Dev** [reply]
Thanks @donalffons,

That's very helpful. Long story short, I don't think I can use `getPrelodadePackage` atm.

I was planning on adding loading percentage to the CascadeStudio "terminal", after looking at the Emscripten docs and the examples and I couldn't get it working, I then checked and the string `getPrelodadePackage` doesn't exist in `/dist/opencascade.wasm.js` which I suspect is because CascadeStudio is on its own fork of opencascade.js.

I might be able to incorporate this in the future though not right now, but I appreciate the advice.
